src/python/pants/backend/scala/target_types.py

```diff
@@ -0,0 +1,39 @@
+# Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+from pants.engine.rules import collect_rules
+from pants.engine.target import COMMON_TARGET_FIELDS
+from pants.jvm.target_types import (
+    JvmArtifact,
+    JvmArtifactArtifactField,
+    JvmArtifactGroupField,
+    JvmArtifactVersionField,
+)
+
+# -----------------------------------------------------------------------------------------------
+# `scala_artifact` target type
+# -----------------------------------------------------------------------------------------------
+
+
+class ScalaArtifactArtifactField(JvmArtifactArtifactField):
+    # TODO: Find a way to have `compute_value` append value of `ScalaSubsystem.version`. Or maybe make
+    #  JvmArtifactArtifactField into an async field?
```

Comment:
@stuhood  / @Eric-Arellano: How do I make this field able to use `ScalaSubsystem.version` in its computation? 

Should I have a UnionRule for resolving artifact names that JVM and Scala could plug into? Then a simple rule would suffice to append the version suffix since the ScalaSubsystem could be a param to the rule.

Reply:
I think that that is what `AsyncField` is for? To allow for a computed field value?

Reply:
Probably simpler is this approach:

https://github.com/pantsbuild/pants/blob/b95ab0e895ec2492eec2208f2a7fe1e783fbd0fb/src/python/pants/backend/python/target_types.py#L443-L454

Reply:
Oh but hm you don't want to have java_artifact need to call `value_or_global_default`, I see. So, two options:

1) Make `JvmArtifactArtifactField` be an async field, so you should call a rule to resolve its value rather than accessing `.value`. This requires lots of new boilerplate and I'm not clear how Scala would hook into that generic rule
2) Use target generation. 

Fundamentally, `Field.compute_value()` cannot access subsystems and changing that would majorly disrupt the Target API.

Reply:
Could a `UnionRule` be used with the types being the target type?

Reply:
What do you mean?

Reply:
The `package` goal uses `FieldSet` classes registered via `UnionRule` to determine what to do with a target. Why not have the JVM and Scala plugins register the target class via `UnionRule` and then the generic rule could do an isinstance check and then request a conversion rule to produce a `ResolvedArtifactCoordinate` type?

Reply:
Oh, yeah - that would essentially be option 1 of using an async field I think.

> I'm not clear how Scala would hook into that generic rule

I think that's what you're saying re UnionRule. You'll have a concrete output type: `ResolvedArtifactCoordinate`, and a union base like `ResolveArtifactCoordinate` with a rule for `jvm_artifact` and another for `scala_artifact`.

Lots of boilerplate, but does that make sense?

Note a limitation: `./pants peek :scala_artifact` will not include the scala version. The only way to do that fwict is using target gen and inspecting the generated `jvm_artifact` target.

Reply:
What's the hesitation with target gen? Personally, I don't like the idea of having _two_ addresses for each scala artifact, but Stu's idea at https://docs.google.com/document/d/17XvhHlx8jBn4K9qFRZPe7qaKJQjTfiKMKYRubnzWxjA/edit fixes that. 

The main weird edge imo is that the address will be `dir#scala_artifact` rather than `dir:whatever_name_i_set_explicitly`. Maybe that's actually better..? We're essentially auto-generating the `name` for you.

Reply:
> What's the hesitation with target gen? 

No hesitation since I apparently didn't even consider it as a solution because my mind had gone right to a Target API solution.

Reply:
> The main weird edge imo is that the address will be `dir#scala_artifact` rather than `dir:whatever_name_i_set_explicitly`. Maybe that's actually better..? We're essentially auto-generating the name for you.

It would be inconsistent with how `jvm_artifact` operates though. Users would need to learn the `#` generator syntax just for this purpose.

Reply:
Agreed, not a fan of that part. If that's the case, I think this is the option:

> Make JvmArtifactField be an async field, so you should call a rule to resolve its value rather than accessing .value. 
>
> ...I think that's what you're saying re UnionRule. You'll have a concrete output type: ResolvedArtifactCoordinate, and a union base like ResolveArtifactCoordinate with a rule for jvm_artifact and another for scala_artifact.